README.md
![GitHub release (latest by date)](https://img.shields.io/github/v/release/carsopre/mistela-flask)
![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/carsopre/mistela-flask)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

# mistela-flask
A multi-event planner elaborated with Flask.

Initial steps done following this [Digital Ocean guide](https://www.digitalocean.com/community/tutorials/how-to-add-authentication-to-your-app-with-flask-login#step-7-setting-up-the-authorization-function)

This version of `Flask` aims to provide a simpile solution to manage an event such as a wedding with different 'sub-events'. The tool also provides a custom-defined admin page as `Flask` does not really provide one itself (and I did not like `Flask-admin` as all users seemed to have admin rights).

Either way, this was a nice user-case for me to get hands-on with Flask.

## Why the name Mistela?
Mistela is 'sort of a wine' quite common in my region (València, Spain) for celebrations or simply after a family dinner. It made sense that if this flask 'flavour' is about events (celebrations) management, it should have an appropiate name, therefore, mistela-flask :).

## Downloading the package.
* Directly from the source:
```cli
pip install git+https://github.com/Carsopre/mistela-flask.git
```
* Or from pypi:
```cli
pip install mistela-flask
```

## Installing the repository.
To develop on the repository you should be using `Poetry`. Once installed, simply run the install command`poetry install` and all development and production dependencies will be added to your virtual environment.

## Usage
Mistela-flask can be used as a package and easily deployed as any regular flask app.
```python
from mistelaflask import create_app
import secrets
import os
os.environ["SECRET_KEY"] = secrets.token_hex(16) # Required environment variable.
os.environ["DATABASE_URI"] = "sqlite:///db.sqlite"
os.environ["MISTELA_TITLE"] = "My big event"
os.environ["STATIC_FOLDER"] = "path//to//my//static//folder"
os.environ["MAIL_USERNAME"] = "joe.doe@email.com"
os.environ["MAIL_PASSWORD"] = "1234"

app = create_app() 
app.run()
```
> By default, mistelaflask will run on a `SQLite` database.

A more 'real' example can be found in the root of the repository as `wsgi.py`.




docs/changelog.md
## v0.16.3 (2022-12-16)

### Fix

- **guest_view_invitations.py**: It is now no longer possible to submit an invitation without responding to the main question

## v0.16.2 (2022-12-16)

### Fix

- **admin_invitation_edit_bulk.html**: Fix updating bulk changes, all fields were returning the baby n value

## v0.16.1 (2022-12-13)

### Fix

- **templates/admin.html**: Fixed burger menu in admin page not showing on phone view

## v0.16.0 (2022-12-12)

### Feat

- **index.html**: Added links to the timeline and invitation to avoid confussion

## v0.15.1 (2022-12-12)

### Fix

- **admin_main_events_list**: Added missing column name

## v0.15.0 (2022-12-12)

### Feat

- **guest_event_extra**: Added extra template that allows to inject html code into the event tab through the MainEvent contact and description fields

## v0.14.7 (2022-12-11)

### Fix

- **guest_event_location.html**: Location -description also parses html code now

## v0.14.6 (2022-12-11)

### Fix

- **templates/events**: Replaced way of adding icons so that more animations are available

## v0.14.5 (2022-12-11)

### Fix

- **templates/event**: it is now possible to add html code into an event description and properly format it with jinja safe

## v0.14.4 (2022-12-11)

### Fix

- **admin_view_guests**: Fixed wrong field during UserEventInvitation

## v0.14.3 (2022-12-11)

### Fix

- **admin_view_main_events.py**: Fix updating location from a main event

## v0.14.2 (2022-12-11)

### Fix

- **admin/events**: Now we can manage an event time (Start / duration)

## v0.14.1 (2022-12-11)

### Fix

- **admin_view_guests.py**: Fix removing user

## v0.14.0 (2022-12-11)

## v0.7.0 (2022-11-20)

### Feat

- **GuestViewEvents**: Added view for events (timeline)
- **guest_invitations**: Guests handle their invitation as 'one' instead of per event
- **guest_invitations.html**: Guest can now manage their own invitations
- **admin_invitation_edit_bulk**: It is now possible to do a bulk edit from the admin menu
- **admin_invitations_summary.html**: Added summary for invitations
- **admin_views/invitations**: Added templates for create, read, delete invitations
- **admin_required**: Added method admin_required to replace and enforce a better decorator
- **admin/guests**: We can now do full CRUD operations on Guests (users without admin rights)
- **admin/guests**: We can now display an admin menu for listing all registered guests and their invitations
- **admin_view_guests.py**: Created new admin view guests file and class
- **admin_events**: Added better admin page for events via tabs
- **auth.py**: Users can now do a first login with a OTP
- **flask_admin**: Added flask admin layer to better handle the models CRUD panel

### Fix

- **auth.py**: Fix mismatch between user and username

## v0.4.0 (2022-11-13)

### Feat

- **mistelaflask/__init__.py**: SECRET_KEY and DATABASE_URI are now read from the environment variables

### Fix

- **pyproject.toml**: Forgot a closing statement

## v0.13.0 (2022-12-11)

### Feat

- **scripts**: We now have available scripts to process images while sandboxing

## v0.12.0 (2022-11-28)

### Feat

- **static**: We now provide the static files through flask environment variables

## v0.11.0 (2022-11-28)

### Feat

- **flask-mail**: Added support to send mails when a user has sent their RSVP

### Fix

- **auth.py**: Fixed logging in
- **auth.py**: Small bug fix when no password still present

## v0.10.0 (2022-11-21)

### Feat

- **guest/events**: Added location information to the event tab

### Fix

- **guest_view_events.py**: Fixed error when no main events present
- **guest_views**: Fixed error when no main events available

## v0.9.0 (2022-11-21)

### Feat

- **locations**: Added  html templates for locations admin view
- **AdminViewLocations**: Created view for admin locations
- **MainEvent**: It is now possible to encapsulate all events under a main one and display it in the index page
- **models**: It is possible to add a new main event with a concrete location for the index view

## v0.8.1 (2022-11-21)

### Fix

- **wsgi.py**: Removed self from local method
- **wsgi.py**: We can now install the development environment when provided

## v0.8.0 (2022-11-21)

### Feat

- **guest_events.html**: Added detailed timeline

## v0.7.0 (2022-11-20)

### Feat

- **GuestViewEvents**: Added view for events (timeline)
- **guest_invitations**: Guests handle their invitation as 'one' instead of per event
- **guest_invitations.html**: Guest can now manage their own invitations

### Fix

- **auth.py**: Fix mismatch between user and username

## v0.5.0 (2022-11-16)

### Feat

- **admin_invitation_edit_bulk**: It is now possible to do a bulk edit from the admin menu
- **admin_invitations_summary.html**: Added summary for invitations
- **admin_views/invitations**: Added templates for create, read, delete invitations
- **admin_required**: Added method admin_required to replace and enforce a better decorator
- **admin/guests**: We can now do full CRUD operations on Guests (users without admin rights)
- **admin/guests**: We can now display an admin menu for listing all registered guests and their invitations
- **admin_view_guests.py**: Created new admin view guests file and class
- **admin_events**: Added better admin page for events via tabs
- **auth.py**: Users can now do a first login with a OTP
- **flask_admin**: Added flask admin layer to better handle the models CRUD panel

## v0.4.0 (2022-11-13)

### Feat

- **mistelaflask/__init__.py**: SECRET_KEY and DATABASE_URI are now read from the environment variables

### Fix

- **pyproject.toml**: Forgot a closing statement

## v0.3.0 (2022-11-13)

### Feat

- **main.py;guest_responses.html**: We now allow the user to update their responses to invitations
- **guest_responses.html**: Added page for a guest to manage their invitations
- **admin/responses**: Added endpoint for admin responses
- **admin.py;admin_guests.html**: Added logic to add guests and display them in order

wsgi.py
import os
import secrets
from datetime import datetime
from pathlib import Path

# Work with the development environment
# add your project directory to the sys.path
_dev_mistela_app = Path(__file__).parent / "mistelaflask"
if _dev_mistela_app.is_dir():
    import sys

    if _dev_mistela_app not in sys.path:
        sys.path = [_dev_mistela_app] + sys.path

from mistelaflask import Flask, create_app, db, models


def init_mistelaflask():
    def init_test_db() -> Flask:
        _app = create_app()
        if "sqlite" in os.environ["DATABASE_URI"]:
            _db_file = Path(__file__).parent / "instance" / "db.sqlite"
            if not _db_file.is_file():
                with _app.app_context():
                    db.create_all()
                    admin = models.User(
                        username="admin",
                        otp="1234",
                        admin=True,
                    )
                    db.session.add(admin)
                    db.session.commit()
            return _app

    os.environ["SECRET_KEY"] = secrets.token_hex(16)  # Required environment variable.
    os.environ["DATABASE_URI"] = "sqlite:///db.sqlite"
    os.environ["MISTELA_TITLE"] = "My big event"
    os.environ["STATIC_FOLDER"] = "path//to//my//static//folder"
    os.environ["MAIL_USERNAME"] = "joe.doe@email.com"
    os.environ["MAIL_PASSWORD"] = "1234"
    return init_test_db()


app = init_mistelaflask()
